### PR TITLE
Adding the bad wkc sms state to respond to jsd-level wkc issues

### PR DESCRIPTION
**Pull request by preston-rogers**
This MR introduces sets all devices to a faulted state when a bad working counter is experienced. This simply uses pre-existing boolean fields in the jsd_t struct

### CODE REVIEW

File: src/jsd/actuator.h [hunk 1]
```diff
@@ -127,6 +127,7 @@ class Actuator : public JsdDeviceBase
     double      high_pos_cmd_limit_eu         = 0.0;
     double      high_pos_cal_limit_eu         = 0.0;
     double      holding_duration_sec          = 0.0;
+    double      wkc_recovery_timeout_sec      = 0.0;
```
Comment (brinkap) [inline src/jsd/actuator.h:130]: This is probably leftover and should be removed

File: src/jsd/actuator.h [hunk 2]
```diff
@@ -206,6 +208,7 @@ class Actuator : public JsdDeviceBase
 
   bool      IsIdleFaultConditionMet();
   FaultType ProcessFaulted();
+  FaultType ProcessBadWkc();
```
Comment (brinkap) [inline src/jsd/actuator.h:211]: This is probably leftover and should be removed

File: src/jsd/jsd_device_base.h [hunk 3]
```diff
@@ -32,6 +32,7 @@ class JsdDeviceBase : public DeviceBase
   void     RegisterSdoResponseQueue(
           std::shared_ptr<std::queue<SdoResponse>> sdo_response_queue);
   void SetOffline(bool is_offline);
+  bool SetBadWkc();
```
Comment (brinkap) [inline src/jsd/jsd_device_base.h:35]: There's no implementation for this method, probably remove this

File: src/manager.cc [hunk 4]
```diff
@@ -224,7 +224,13 @@ bool fastcat::Manager::Process(double external_time)
...
   std::lock_guard<std::mutex> lock(parameter_mutex_);
   for (auto it = jsd_map_.begin(); it != jsd_map_.end(); ++it) {
-    jsd_read(it->second, 1e6 / target_loop_rate_hz_);
+    auto ifname = it->first;
+    auto jsd = it->second;
+    jsd_read(jsd, 1e6 / target_loop_rate_hz_);
+    if (jsd->wkc != jsd->expected_wkc && !IsFaulted()) {
+      ERROR("Bad working counter experienced on jsd bus %s", ifname.c_str());
+      ExecuteAllDeviceFaults();
+    }
```
Comment (brinkap) [inline src/manager.cc:233]: Nice! simple and effective solution to WKC faults. I'm curious to hear how this pans out with testing!

**Review by brinkap — CHANGES_REQUESTED** [review]
There are a few minor leftover changes related to the reverted actuator FSM changes, otherwise I like the simple, clean bus fault response.

File: src/manager.h [hunk 5]
```diff
@@ -230,7 +230,7 @@ class Manager
 
   double                        target_loop_rate_hz_                = 0.0;
   bool                          zero_latency_required_              = true;
-  bool                          faulted_                            = false;
+  bool                          faulted_                            = true;
```
Comment (alex-brinkman) [inline src/manager.h:233]: Why start the manager faulted?

File: src/manager.cc [hunk 6]
```diff
@@ -224,7 +224,14 @@ bool fastcat::Manager::Process(double external_time)
...
   for (auto it = jsd_map_.begin(); it != jsd_map_.end(); ++it) {
-    jsd_read(it->second, 1e6 / target_loop_rate_hz_);
+    auto ifname = it->first;
+    auto jsd = it->second;
+    jsd_read(jsd,  1e6 / target_loop_rate_hz_);
+    if (jsd->wkc != jsd->expected_wkc && !IsFaulted()) {
+      ERROR("Bad working counter experienced on jsd bus %s", ifname.c_str());
+      jsd_inspect_context(jsd);
+      ExecuteAllDeviceFaults();
+    }
```
Comment (alex-brinkman) [inline src/manager.cc:234]: Great!

File: CMakeLists.txt [hunk 7]
```diff
@@ -74,7 +74,7 @@ find_package(YamlCpp REQUIRED 0.6.3)
 include(FetchContent)
 FetchContent_Declare(jsd
     GIT_REPOSITORY https://github.com/nasa-jpl/jsd.git
-    GIT_TAG v3.0.1
+    GIT_TAG prestonr-feature-add-wkc-driver-saves
```
Comment (alex-brinkman) [inline CMakeLists.txt:77]: Pin this to the fixed JSD release before finishing the MR

**Review by alex-brinkman — CHANGES_REQUESTED** [review]
This is a readable, clean solution - I approve. Please take care of the 2 minor items before you merge

File: src/manager.h (hunk 5 above)
Comment (preston-rogers) [reply src/manager.h:233]: The manager calls Process() and then the current wkc is checked against the expected wkc. If these differ and IsFaulted() returns false, we fault all devices. This makes sense during nominal operation but when ConfigFromYaml() is called Process() is called twice and I found the system would immediately error out due to wkc issues with my implementation.
Starting the system in faulted state causes the wkc check to be ignored and allows the two Process() calls to occur before ExecuteAllDeviceResets() is called (at the end of the ConfigFromYaml() method).
I found this to be the simplest change to get around this issue.

**Pull request merged**